subgraphs/truefi/src/entities/market.ts

```diff
@@ -0,0 +1,464 @@
+  hourlySnapshot.hourlyTotalRevenueUSD =
+    hourlySnapshot.hourlyTotalRevenueUSD.plus(revenueAmountUSD);
+  hourlySnapshot.save();
+  addProtocolSideRevenue(event, revenueAmountUSD);
+}
+
+export function openMarketBorrowerPosition(market: Market): void {
+  market.openPositionCount += 1;
+  market.positionCount += 1;
+  market.borrowingPositionCount += 1;
+  market.save();
+  incrementProtocolPositionCount();
+}
+
+export function openMarketLenderPosition(market: Market): void {
+  market.openPositionCount += 1;
+  market.positionCount += 1;
+  market.lendingPositionCount += 1;
+  market.save();
+  incrementProtocolPositionCount();
+}
+
+export function closeMarketPosition(market: Market): void {
+  market.openPositionCount -= 1;
+  market.closedPositionCount += 1;
+  market.save();
+  decrementProtocolOpenPositionCount();
+}
+
+function updateMarketTVL(
+  event: ethereum.Event,
+  market: Market,
+  isPool: bool
+): void {
+  let inputTokenPriceUSD = market.inputTokenPriceUSD;
+  let totalValueLockedUSD: BigDecimal = BIGDECIMAL_NEGATIVE_ONE;
+  const inputToken = getTokenById(market.inputToken);
+
+  if (isPool) {
+    const poolContract = TruefiPool2.bind(Address.fromString(market.id));
+    const currencyBalance = poolContract.try_currencyBalance();
+    if (!currencyBalance.reverted) {
+      market.inputTokenBalance = currencyBalance.value;
+    }
+    const totalSupply = poolContract.try_totalSupply();
+    if (!totalSupply.reverted) {
+      market.outputTokenSupply = totalSupply.value;
+    }
+    const poolValue = poolContract.try_poolValue();
+    if (!poolValue.reverted) {
+      totalValueLockedUSD = bigIntToBigDecimal(
+        poolValue.value,
+        inputToken.decimals
+      ).times(inputTokenPriceUSD);
+    }
+  } else {
+    const portfolioContract = ManagedPortfolio.bind(
+      Address.fromString(market.id)
+    );
+    const liquidValue = portfolioContract.try_liquidValue();
+    if (!liquidValue.reverted) {
+      market.inputTokenBalance = liquidValue.value;
+    }
+    const totalSupply = portfolioContract.try_totalSupply();
+    if (!totalSupply.reverted) {
+      market.outputTokenSupply = totalSupply.value;
+    }
+    const portfolioValue = portfolioContract.try_value();
+    if (!portfolioValue.reverted) {
+      totalValueLockedUSD = bigIntToBigDecimal(
+        portfolioValue.value,
+        inputToken.decimals
+      ).times(inputTokenPriceUSD);
+    }
+  }
+
+  if (totalValueLockedUSD.ge(BIGDECIMAL_ZERO)) {
+    const changeUSD = totalValueLockedUSD.minus(market.totalValueLockedUSD);
+    updateProtocolTVL(event, changeUSD, changeUSD);
+    if (market.outputTokenSupply.gt(BIGINT_ZERO)) {
+      market.outputTokenPriceUSD = totalValueLockedUSD.div(
+        bigIntToBigDecimal(
+          market.outputTokenSupply,
+          getTokenById(market.outputToken!).decimals
+        )
+      );
+    }
+    market.totalValueLockedUSD = totalValueLockedUSD;
+    market.totalDepositBalanceUSD = totalValueLockedUSD;
```

Comment:
Actually it is saved after this updateMarketTVL() function is returned. I want to follow the existing pattern in the codes before and I find in many cases the entity is not saved in the function, but after the function calling. It may be good if there are many consecutive function call. I don't know which way is a better one.

Reply:
I didn't see that, but take a look at this and lmk what you think. Sometimes issues have popped up with saving / not saving. and it shouldn't really hurt perf to save more often

https://github.com/messari/subgraphs/blob/master/docs/Mapping.md#saving-entities

Reply:
Added in a955c4539656c927650be52eaf6dad04e7bfc825 Thanks!